app-core/src/main/kotlin/com/flab/hsw/core/domain/user/usecase/UserLoginUseCase.kt

```diff
@@ -0,0 +1,46 @@
+package com.flab.hsw.core.domain.user.usecase
+
+import com.flab.hsw.core.annotation.UseCase
+import com.flab.hsw.core.domain.user.User
+import com.flab.hsw.core.domain.user.aggregate.PasswordEncryptor
+import com.flab.hsw.core.domain.user.exception.InvalidUserPasswordException
+import com.flab.hsw.core.domain.user.exception.UserByLoginIDNotFoundException
+import com.flab.hsw.core.domain.user.repository.UserRepository
+
+interface UserLoginUseCase {
+
+    fun loginProcess(message: UserLoginMessage): User
+
+    interface UserLoginMessage {
+        val loginId: String
+        val password: String
+    }
+
+    companion object {
+        fun newInstance(
+            userRepository: UserRepository,
+            passwordEncryptor: PasswordEncryptor = PasswordEncryptor.newInstance()
+        ): UserLoginUseCase = UserLoginUseCaseImpl(
+            userRepository,
+            passwordEncryptor
+        )
+    }
+
+}
+
+@UseCase
+internal class UserLoginUseCaseImpl(
+    private val userRepository: UserRepository,
+    private val passwordEncryptor: PasswordEncryptor
+) : UserLoginUseCase {
+    override fun loginProcess(message: UserLoginUseCase.UserLoginMessage): User {
+        val findUser = userRepository.findByLoginId(message.loginId)
+            ?: throw UserByLoginIDNotFoundException(message.loginId)
+
+        findUser.updateLastActiveTimeToNow()
+        userRepository.update(findUser)
+
+        return if(passwordEncryptor.isMatched(message.password, findUser.password)) findUser
```

Comment:
여기서 password match 가 일어나지 않았을 때, `updateLastActiveTimeToNow` 를 되돌리려면 어떻게 하는게 좋을까요,

또는,

검사를 먼저 한 다음, 옳바른 이용자에게만 `updateLastActiveTimeToNow` 를 호출하는게 나을까요?

Reply:
그리고 한번 로그인한 이용자의 session 은 언제 만료되나요?

Reply:
kotlin stdlib 의 `also` 를 쓰면 좀더 풍부한 표현을 하실 수 있는 사례같습니다.

Reply:
> 한번 로그인한 이용자의 session 은 언제 만료되나요?

Session Timout은 **WAS(Tomcat) default 설정인 30분** 입니다.

PR의 Key Implement에 남겼던 내용인데요. 비밀번호 확인을 진행한 뒤 `app-main`에서 WAS Session에 (로그인 완료된)반환된 객체를 저장해두는 방식으로 진행하려 합니다. 

이 경우 `app-core` 모듈에는 세션 상태가 표현되지 않는데 올바른 걸까요,,? 이전에 `app-core`만 실행해도 애플리케이션이 정상 동작해야 한다고 말씀하셨던게 기억나서 질문드립니다.

Reply:
> kotlin stdlib 의 `also` 를 쓰면 좀더 풍부한 표현을 하실 수 있는 사례같습니다.

[Scope Function 'also' - Kotlin Docs](https://kotlinlang.org/docs/scope-functions.html#also)에서 `also` 함수의 좋은 용례를 소개하고 있네요.

> *`also` is good for performing some actions that take the context object as an argument. Use `also` for actions that need a reference to the object rather than its properties and functions, or when you don't want to shadow the this reference from an outer scope.*
>
> *`also`는 컨텍스트 객체를 **인자로 사용하는 작업을 수행**하는 경우 유용합니다. 객체의 프로퍼티나 함수가 아닌 **객체의 참조가 필요한 작업** 혹은 **외부(scope 외부)로부터 객체의 참조를 숨기고 싶은 경우** 유용합니다.*

위 내용을 고려해 수정해보겠습니다. 

```kotlin
if(passwordEncryptor.isMatched(message.password, findUser.password)){
    throw InvalidUserPasswordException()
}

return findUser.apply {
    this.updateLastActiveTimeToNow()  // 객체의 참조가 필요한 작업
    userRepository.update(this)  // 인자로 사용하는 작업
}
```